### PR TITLE
Improve stateless validation

### PR DESCRIPTION
**Pull request by nickaleks**
### Description of the Change

This Pull Request introduces several changes to stateless validation:
1. FieldValidator now returns value of invalid field
2. Amount is checked to be greater than 0
3. CommandsValidator now appends counter to Command name, which allows understanding which command has failed.

The biggest change of this pull request is tests for FieldValidator, which are organized in a table-like manner with various test cases for different fields. Because of a lot of boilerplate, file is quite large, but it allows to add various test cases for different fields very easily.

### Benefits

1. Better error message for stateless validation
2. Test structure for FieldValidator allows us to test it easier

### Possible Drawbacks 

1. Size of the test file is quite large, and it may take some time to understand how it all works. 

### Usage Examples or Tests *[optional]*

If you want to add a new constraint to some field, you can test it by adding a test case to appropriate map in file field_validator_test.cpp.

### Alternate Designs *[optional]*
Google Test has parametrized test which allows running a test with different values however it would complicate code even, and require even more boilerplate code.

### CODE REVIEW

File: shared_model/validators/field_validator.hpp [hunk 1]
```diff
@@ -48,27 +48,37 @@ namespace shared_model {
...
+        }
       }
 
       void validatePubkey(ReasonsGroupType &reason,
                           const interface::types::PubkeyType &pubkey) const {
-        if (pubkey.blob().size() != 32) {
-          reason.second.push_back("Public key has wrong size");
+        if (pubkey.blob().size() != key_size) {
+          auto message =
+              generateErrorMessage("Public key has wrong size",
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: *couldhave*
IMO `generateErrorMessage` can be just alias for `boost::format`. It is not very good to use `std::to_string`. What if I, as a developer of validator, want to make some comprehensive error? Like, `str+number+str+hex+number` (I hope you get the idea)

File: shared_model/validators/field_validator.hpp [hunk 2]
```diff
@@ -48,27 +48,37 @@ namespace shared_model {
           ReasonsGroupType &reason,
           const interface::types::AccountIdType &account_id) const {
         if (not std::regex_match(account_id, account_id_)) {
-          reason.second.push_back("Wrongly formed account_id");
+          auto message =
+              generateErrorMessage("Wrongly formed account_id",
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: *couldhave*
As for error message, it would be good to return matching regex to the user in error message, so the user will be able to fix his code without having to go to the documentation.

File: shared_model/validators/field_validator.hpp [hunk 3]
```diff
@@ -77,46 +87,60 @@ namespace shared_model {
           const interface::AddPeer::AddressType &address) const {
         if (not(iroha::validator::isValidIpV4(address)
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: Are you sure that validation should look like this? Because, as I remember, `address` here is `ip:port`, not just address.
If this checks only for IPv4 or hostname compliance, then these will be failed, but should not:
```
1.1.1.1:50
domain.com:50
```

File: shared_model/validators/field_validator.hpp [hunk 4]
```diff
@@ -48,27 +48,37 @@ namespace shared_model {
           ReasonsGroupType &reason,
           const interface::types::AccountIdType &account_id) const {
         if (not std::regex_match(account_id, account_id_)) {
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: *musthave*

`account_id_` is bad name for regex. 

I understand that this is not part of this PR, but please, move all regexp out of constructor 
 (https://github.com/hyperledger/iroha/blob/67a82efcb776283f85a6b912f89d86cbba0a9837/shared_model/validators/field_validator.hpp#L39), make them as static members and give them meaningful names:
```
account_id_regexp
...
```

File: shared_model/validators/field_validator.hpp [hunk 5]
```diff
@@ -160,33 +187,42 @@ namespace shared_model {
...
+
+        auto time_message =
+            (boost::format("%llu, now: %llu") % timestamp % now).str();
         if (now < timestamp) {
-          reason.second.push_back(boost::str(
-              boost::format(
-                  "timestamp broken: send from future (%llu, now %llu)")
-              % timestamp % now));
+          auto message = generateErrorMessage(
+              "timestamp broken: sent from future", time_message);
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: **bad timestamp**: sent from future

File: shared_model/validators/field_validator.hpp [hunk 6]
```diff
@@ -160,33 +187,42 @@ namespace shared_model {
...
+          reason.second.push_back(std::move(message));
         }
 
-        if (now - timestamp > MAX_DELAY) {
-          reason.second.push_back(boost::str(
-              boost::format("timestamp broken: too old (%llu, now %llu)")
-              % timestamp % now));
+        if (now - timestamp > max_delay) {
+          auto message =
+              generateErrorMessage("timestamp broken: too old", time_message);
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: **bad timestamp**: too old

File: shared_model/validators/field_validator.hpp [hunk 7]
```diff
@@ -160,33 +187,42 @@ namespace shared_model {
...
      private:
       std::regex account_id_, asset_id_, ip_address_, name_, detail_key_;
       // max-delay between tx creation and validation
-      static constexpr auto MAX_DELAY =
+      static constexpr auto max_delay =
           std::chrono::hours(24) / std::chrono::milliseconds(1);
+      // size of key
+      static constexpr auto key_size = 32;
+
+      std::string generateErrorMessage(const std::string &error,
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: Now I am thinking that this method is not needed at all, thus making a developer who writes a validator responsible for the message.

File: shared_model/validators/transaction_validator.hpp [hunk 8]
```diff
@@ -189,7 +189,8 @@ namespace shared_model {
       ReasonsGroupType operator()(
           const detail::PolymorphicWrapper<interface::SetQuorum> &sq) const {
         ReasonsGroupType reason;
-        reason.first = "SetQuorum";
+        addInvalidCommand(reason, "SetAccountQuorum");
```
Comment (Warchant) [inline shared_model/validators/transaction_validator.hpp]: SetQuorum or SetAccountQuorum?

Command's name is SetQuorum

File: shared_model/validators/transaction_validator.hpp [hunk 9]
```diff
@@ -226,6 +227,15 @@ namespace shared_model {
 
      private:
       FieldValidator validator_;
+      mutable int command_counter{0};
```
Comment (Warchant) [inline shared_model/validators/transaction_validator.hpp]: why `mutable`?

File: shared_model/validators/field_validator.hpp (hunk 2 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Isn't it insecure to expose regexes?

File: shared_model/validators/field_validator.hpp [hunk 10]
```diff
@@ -48,27 +48,37 @@ namespace shared_model {
...
+          auto message = generateErrorMessage(
+              "Wrongly formed asset_id", static_cast<std::string>(asset_id));
+          reason.second.push_back(std::move(message));
         }
       }
 
       void validateAmount(ReasonsGroupType &reason,
                           const interface::Amount &amount) const {
-        /* put here any validations*/
+        if (amount.intValue() == 0) {
```
[the remark this answers is not in the document]
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: It is unsigned, but I agree to change the condition to > 0 and do that for tx_counter as well because it is clearer and I had the same question when I saw this code for tx_counter.

File: shared_model/validators/field_validator.hpp (hunk 3 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: address, in this case, will contain ip:port, or hostname. You can check out test/module/libs/validator,  there are tests for this functions

File: shared_model/validators/transaction_validator.hpp (hunk 8 above)
Comment (nickaleks) [reply shared_model/validators/transaction_validator.hpp]: Typo

File: shared_model/validators/field_validator.hpp (hunk 2 above)
Comment (Warchant) [reply shared_model/validators/field_validator.hpp]: 1. Software is opensource, users are able to look at them by themselves
2. It is better for UX

File: shared_model/validators/transaction_validator.hpp (hunk 9 above)
Comment (nickaleks) [reply shared_model/validators/transaction_validator.hpp]: Because all validation methods are const, and we want to modify counter in them.

File: shared_model/validators/field_validator.hpp (hunk 7 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: I had similar thoughts, it just makes everything more complicated, so I'll remove it.

File: shared_model/validators/field_validator.hpp (hunk 2 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Okay, will add.

File: shared_model/validators/field_validator.hpp (hunk 4 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Isn't it a bit unnecessary to add 'regexp' to a variable name, since its type is already std::regex?

File: shared_model/validators/field_validator.hpp (hunk 1 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Why is std::to_string bad?

File: shared_model/validators/field_validator.hpp (hunk 4 above)
Comment (Warchant) [reply shared_model/validators/field_validator.hpp]: @nickaleks it is obvious for you in IDE, but when others read the code, it is quite hard to understand just by looking at the variable.

Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Okay, also I think since this class does not have any state, we can as well use free functions in a proper namespace, as already done for example in address validation. But it is debatable because we tend to use classes everywhere.

Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Also, this change would require its own pull request because right now if we declare regexes static, we have to define them after class declaration. So I will fix it when we separate declaration and definition of FieldValidator.

File: shared_model/validators/field_validator.hpp (hunk 2 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Sadly, there seems to be no way to extract a pattern from a string, so this change would require putting pattern in a separate variable. I would do that in separate pull request along with separation of .hpp and .cpp file.

**Comment by Warchant** [comment]
However, could you also fix minor issues from sorabot?

File: shared_model/validators/field_validator.hpp (hunk 4 above)
Comment (lebdron) [reply shared_model/validators/field_validator.hpp]: @Warchant `std::regex` has a limited set of operations, so that if the variable is referenced in the code with `regex_match` for example, it is clear for the reader that referenced variable is indeed a regex.

File: shared_model/validators/field_validator.hpp [hunk 11]
```diff
@@ -36,39 +36,50 @@ namespace shared_model {
...
+                  .str();
+          reason.second.push_back(std::move(message));
         }
       }
 
       void validateAmount(ReasonsGroupType &reason,
                           const interface::Amount &amount) const {
-        /* put here any validations*/
+        if (amount.intValue() <= 0) {
+          reason.second.push_back("Amount must be greater than 0");
```
Comment (lebdron) [inline shared_model/validators/field_validator.hpp]: It might be useful to print out received value just as In other validators, for the same reason.

File: shared_model/validators/field_validator.hpp [hunk 12]
```diff
@@ -160,33 +200,39 @@ namespace shared_model {
...
-          reason.second.push_back(
-              boost::str(boost::format("Counter should be > 0")));
+        if (counter <= 0) {
+          reason.second.push_back("Counter should be > 0");
         }
       }
 
      private:
-      std::regex account_id_, asset_id_, ip_address_, name_, detail_key_;
+      std::regex account_id_regex, asset_id_regex, name_regex, detail_key_regex;
```
Comment (lebdron) [inline shared_model/validators/field_validator.hpp]: https://google.github.io/styleguide/cppguide.html#Variable_Names

>Data members of classes, both static and non-static, are named like ordinary nonmember variables, but with a trailing underscore.

File: shared_model/validators/field_validator.hpp [hunk 13]
```diff
@@ -160,33 +200,39 @@ namespace shared_model {
...
         }
       }
 
       void validateCounter(ReasonsGroupType &reason,
                            const interface::types::CounterType &counter) const {
-        if (counter == 0) {
-          reason.second.push_back(
-              boost::str(boost::format("Counter should be > 0")));
+        if (counter <= 0) {
+          reason.second.push_back("Counter should be > 0");
```
Comment (lebdron) [inline shared_model/validators/field_validator.hpp]: Value can also be printed out.

File: shared_model/validators/field_validator.hpp [hunk 14]
```diff
@@ -36,39 +36,50 @@ namespace shared_model {
...
 
       void validateAccountId(
           ReasonsGroupType &reason,
           const interface::types::AccountIdType &account_id) const {
-        if (not std::regex_match(account_id, account_id_)) {
-          reason.second.push_back("Wrongly formed account_id");
+        if (not std::regex_match(account_id, account_id_regex)) {
+          auto message =
+              (boost::format("Wrongly formed account_id, passed value: '%s'")
+               % static_cast<std::string>(account_id))
```
Comment (lebdron) [inline shared_model/validators/field_validator.hpp]: It's not clear whether cast is required, since it seems unlikely that `AccountIdType` will have only explicit conversion to `std::string`. Same applies to other methods.

File: shared_model/validators/transaction_validator.hpp [hunk 15]
```diff
@@ -144,7 +144,7 @@ namespace shared_model {
           const detail::PolymorphicWrapper<interface::GrantPermission> &gp)
           const {
         ReasonsGroupType reason;
-        reason.first = "GrantPermission";
+       addInvalidCommand(reason, "GrantPermission");
```
Comment (lebdron) [inline shared_model/validators/transaction_validator.hpp]: Space is missing before `addInvalidCommand`.

File: shared_model/validators/transaction_validator.hpp [hunk 16]
```diff
@@ -189,7 +189,8 @@ namespace shared_model {
       ReasonsGroupType operator()(
           const detail::PolymorphicWrapper<interface::SetQuorum> &sq) const {
         ReasonsGroupType reason;
-        reason.first = "SetQuorum";
+        addInvalidCommand(reason, "SetQuorum");
+
```
Comment (lebdron) [inline shared_model/validators/transaction_validator.hpp]: Newline seems redundant.

File: shared_model/validators/transaction_validator.hpp [hunk 17]
```diff
@@ -201,7 +202,7 @@ namespace shared_model {
           const detail::PolymorphicWrapper<interface::SubtractAssetQuantity>
               &saq) const {
         ReasonsGroupType reason;
-        reason.first = "SubtractAssetQuantity";
+        addInvalidCommand(reason,"SubtractAssetQuantity");
```
Comment (lebdron) [inline shared_model/validators/transaction_validator.hpp]: Space between `,` and `"SubtractAssetQuantity"` is missing. Maybe run clang_format?

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 18]
```diff
@@ -0,0 +1,747 @@
...
+  // possible commands, some fields are checked several times
+  // because they are present in several commands. To prevent that, this set
+  // contains all already tested fields, and checked each time we try to test
+  // a field
+  std::unordered_set<std::string> checked_fields;
+
+  /************************** TEST CASES ***************************/
+
+  std::vector<FieldTestCase> account_id_test_cases{
+      {
```
Comment (lebdron) [inline test/module/shared_model/validators/field_validator_test.cpp]: Replacing this with some sort of function which takes case name, account id, bool (which will in turn create correct expected error message) will improve readability and reduce code duplication. Same applies for code below.

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 19]
```diff
@@ -0,0 +1,747 @@
...
+  std::vector<FieldTestCase> tx_hashes_test_cases;
+
+  /**************************************************************************/
+
+  // register validation function and test cases
+  std::unordered_map<std::string, FieldTest> field_validators{
+      // Command fields
+      {"account_id",
+       {[&] {
+          validation::ReasonsGroupType reason;
```
Comment (lebdron) [inline test/module/shared_model/validators/field_validator_test.cpp]: Also looks like a shortcut is required, so that `reason` lines are not duplicated.

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 20]
```diff
@@ -0,0 +1,747 @@
...
+        return iroha::protocol::Query::Payload::descriptor()->FindOneofByName(
+            "query");
+      },
+      [&](auto field) {
+        // Set concrete type for new query
+        return payload->GetReflection()->MutableMessage(payload, field);
+      },
+      [this](auto field, auto query) { this->runTestCases(field); },
+      [&] {});
+}
```
Comment (lebdron) [inline test/module/shared_model/validators/field_validator_test.cpp]: It looks like test for query fields is missing. The purpose is the same as `TransactionFieldsValidation`, since `QueryFieldsValidation` only does validation of concrete query (such as GetAccount) fields. It could iterate `iroha::protocol::Query::descriptor()`.

Comment (nickaleks) [reply test/module/shared_model/validators/field_validator_test.cpp]: What do you mean by missing? if a field of a query is not covered by tests, an exception will be thrown inside runTestCases()

File: shared_model/validators/field_validator.hpp (hunk 12 above)
Comment (Warchant) [reply shared_model/validators/field_validator.hpp]: And probably it is better to write all strings in separate lines, like:
```
std::regex account_id_regex_;
std::regex asset_id_regex_;
...
```

File: shared_model/validators/transaction_validator.hpp (hunk 17 above)
Comment (Warchant) [reply shared_model/validators/transaction_validator.hpp]: Just use `clang-format`

File: test/module/shared_model/validators/field_validator_test.cpp (hunk 18 above)
Comment (nickaleks) [reply test/module/shared_model/validators/field_validator_test.cpp]: Do you suggest define a function for each field, which allows test case registration, like "registerAccountIDTestcase(name, account id, should_pass, error_message)"?

File: test/module/shared_model/validators/field_validator_test.cpp (hunk 19 above)
Comment (nickaleks) [reply test/module/shared_model/validators/field_validator_test.cpp]: I was thinking about passing a reference to the method to function, but non-static class members cannot be passed that way, I can wrap them in lambda, but the amount of code would not decrease in this case.

File: test/module/shared_model/validators/validators_fixture.hpp [hunk 21]
```diff
@@ -29,8 +29,11 @@
 class ValidatorsTest : public ::testing::Test {
  public:
   ValidatorsTest() {
-    valid_amount.set_precision(2);
-    valid_amount.mutable_value()->set_fourth(1000);
+    amount.set_precision(2);
+    amount.mutable_value()->set_fourth(1000);
+    public_key_size = 32;
+    hash_size = 32;
+    counter = 0;
```
Comment (l4l) [inline test/module/shared_model/validators/validators_fixture.hpp]: Can't find any place where it is used

File: test/module/shared_model/validators/validators_fixture.hpp [hunk 22]
```diff
@@ -117,28 +120,51 @@ class ValidatorsTest : public ::testing::Test {
...
+    account_name = "admin";
+    domain_id = "ru";
+    detail_key = "key";
+    public_key = std::string(public_key_size, '0');
+    hash = std::string(public_key_size, '0');
+    role_permission = iroha::protocol::RolePermission::can_append_role;
+    grantable_permission =
+        iroha::protocol::GrantablePermission::can_add_my_signatory;
+    quorum = 2;
+    precision = 42;
```
Comment (l4l) [inline test/module/shared_model/validators/validators_fixture.hpp]: Is it fine that precision doesn't match with the one at the `ValidatorsTest::ValidatorsTest()`?

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 23]
```diff
@@ -0,0 +1,747 @@
...
+ public:
+  // To test one field, validation function is required,
+  // which is always the same, and several test cases, which represent
+  // various inputs
+  using FieldTest = std::pair<ValidationFunction, std::vector<FieldTestCase>>;
+
+  // Run all test cases for given field
+  void runTestCases(const google::protobuf::FieldDescriptor *field) {
+    auto field_name = field->name();
+    // skip field, if already tested
```
Comment (l4l) [inline test/module/shared_model/validators/field_validator_test.cpp]: What's the problem if it runs twice?

File: shared_model/validators/field_validator.hpp [hunk 24]
```diff
@@ -160,33 +203,46 @@ namespace shared_model {
...
         if (now < timestamp) {
-          reason.second.push_back(boost::str(
-              boost::format(
-                  "timestamp broken: send from future (%llu, now %llu)")
-              % timestamp % now));
+          auto message = (boost::format("bad timestamp: sent from future, "
+                                        "timestamp: %llu, now: %llu")
+                          % timestamp % now)
+                             .str();
+          reason.second.push_back(std::move(message));
```
Comment (Warchant) [inline shared_model/validators/field_validator.hpp]: this branch is not covered by tests

File: shared_model/validators/field_validator.hpp [hunk 25]
```diff
@@ -36,39 +36,54 @@ namespace shared_model {
...
      public:
       FieldValidator()
-          : account_id_(R"([a-z]{1,9}\@[a-z]{1,9})"),
-            asset_id_(R"([a-z]{1,9}\#[a-z]{1,9})"),
-            ip_address_(
-                "((([0-1]?\\d\\d?)|((2[0-4]\\d)|(25[0-5]))).){3}(([0-1]?\\d\\d?"
-                ")|((2[0-4]\\d)|(25[0-5])))"),
-            name_(R"([a-z]{1,9})"),
-            detail_key_(R"([A-Za-z0-9_]{1,})") {}
+          : account_id_pattern_(R"([a-z]{1,9}\@[a-z]{1,9})"),
```
Comment (motxx) [inline shared_model/validators/field_validator.hpp]: account_id validation has changed to RFC1035.
It is difficult to express it by regex. `boost::spirit::qi`(EBNF) is used and the parser is implemented in `validator::isValidDomainName()`
https://github.com/hyperledger/iroha/pull/694

File: shared_model/validators/field_validator.hpp [hunk 26]
```diff
@@ -160,33 +203,46 @@ namespace shared_model {
...
 
      private:
-      std::regex account_id_, asset_id_, ip_address_, name_, detail_key_;
+      std::regex account_id_pattern_;
+      std::regex asset_id_pattern_;
+      std::regex name_pattern_;
+      std::regex detail_key_pattern_;
       // max-delay between tx creation and validation
-      static constexpr auto MAX_DELAY =
+      static constexpr auto max_delay =
```
[the remark this answers is not in the document]
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: AFAIK, ALL_CAPS names are reserved for preprocessor macros

File: test/module/shared_model/validators/field_validator_test.cpp (hunk 23 above)
Comment (nickaleks) [reply test/module/shared_model/validators/field_validator_test.cpp]: Twice - no issue, but it can run as many times as number of commands that have this field. Even if it is fast, I would consider 10x increase in test time to be unwanted.

File: shared_model/validators/field_validator.hpp (hunk 25 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: I suggest we make that a separate pull request.

File: shared_model/validators/field_validator.hpp (hunk 24 above)
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: Right now it is hard to test timestamps, because we cannot verify that message is correct. I will think of a solution and maybe do a separate pull request

File: shared_model/validators/transaction_validator.hpp (hunk 9 above)
Comment (nickaleks) [reply shared_model/validators/transaction_validator.hpp]: It could be used in different functions, and I do not think static variable provides any major benefit over class member, and class member expresses intent more clearly.

File: test/module/irohad/consensus/yac/network_test.cpp [hunk 27]
```diff
@@ -43,13 +42,15 @@ namespace iroha {
 
           grpc::ServerBuilder builder;
           int port = 0;
-          builder.AddListeningPort(peer.address,
+          builder.AddListeningPort("0.0.0.0:0",
```
[the remark this answers is not in the document]
Comment (nickaleks) [reply test/module/irohad/consensus/yac/network_test.cpp]: This change is not part of this PR, I don't know how these commits ended up here, I guess I need to rebase :)

Also I think removing hardcoded literals like this one should be a separate issue, so that they all get organized in one place and used by all tests.

File: test/module/shared_model/validators/validators_fixture.hpp (hunk 21 above)
Comment (nickaleks) [reply test/module/shared_model/validators/validators_fixture.hpp]: Yeah, they are redundant, will remove.

File: shared_model/validators/field_validator.hpp [hunk 28]
```diff
@@ -160,33 +203,46 @@ namespace shared_model {
...
           std::chrono::hours(24) / std::chrono::milliseconds(1);
```
[the remark this answers is not in the document]
Comment (nickaleks) [reply shared_model/validators/field_validator.hpp]: The problem is that we need to add 'using namespace std::chrono_literals;', which is not allowed in header files by sorabot

[the remark this answers is not in the document]
Comment (l4l) [reply shared_model/validators/field_validator.hpp]: Ah, I see and the exact using cause warning as well. Ok then

File: test/module/irohad/consensus/yac/network_test.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (l4l) [reply test/module/irohad/consensus/yac/network_test.cpp]: The problem that you've introduced magic values here. It's better write clean code so there's no need in additional issues for fixing tech debt

File: test/module/shared_model/validators/field_validator_test.cpp (hunk 23 above)
Comment (l4l) [reply test/module/shared_model/validators/field_validator_test.cpp]: Looks quite hackish, but let it be. Hopefully that can save time on debugging in future

File: shared_model/validators/transaction_validator.hpp (hunk 9 above)
Comment (l4l) [reply shared_model/validators/transaction_validator.hpp]: It's possible to encapsulate this var in separate function for that and I don't see any issues here. Moreover it can be helpful in enhancing with the async support

File: test/module/shared_model/validators/field_validator_test.cpp (hunk 23 above)
Comment (nickaleks) [reply test/module/shared_model/validators/field_validator_test.cpp]: This is hackish, because we are limited by protobuf reflection, if we can get rid of it in the future, this code will hopefully go away.
